### ISSUE TITLE
How merge 9 one channel tiffs into one 9 channels tiff 

### ISSUE THREAD

**Issue by gmh1033286494** [libvips/pyvips#223]
Hi, @jcupitt.

I am trying to merge my results into one mutli-channel tiff. I tried arrayjoin() but it seems to put them together into one channel. Instead I want each results to be one channel of my new tiff. Could you give some suggestions?

Thank you very much.

**Comment by jcupitt** [comment]
Hello @gmh1033286494, you need `bandjoin`. For example:

```python
rgb = r.bandjoin([g, b])
```

To join three one-band images to make one three-band image.

**Comment by gmh1033286494** [comment]
Great! It worked.
After I concatenated 2 images,I realized that each images have 3 bands(I think they are RGB, although my inputs and outputs are all black and white.) so there are 6 bands in my result images. I checked with Qupath, the Dimension is 3*1*1. I am wondering if Qupath only shows the result as one whole image even thought there are 6 bands in it.   

My question is 1 How to use pyvips to convert my result to colored RGB. 2 Is there any way I can see the fact that I actually did concatenate several tiff into one tiff?

Thank you very much.

**Comment by jcupitt** [comment]
You can use `tiffinfo` to see exactly what's inside a TIFF file.

If you want to make a 9-band image you can load into QuPath, you need to format it carefully. You have to code up the image as a set of one band images stacked vertically, you need to set `page-height` to the height of each frame, and you have to attach some XML to `image-description` saying what each frame means. It's a bit hairy :(

**Comment by jcupitt** [comment]
Sample code: https://github.com/libvips/pyvips/issues/222#issuecomment-748509935

**Comment by gmh1033286494** [comment]
> You can use `tiffinfo` to see exactly what's inside a TIFF file.
> 
> If you want to make a 9-band image you can load into QuPath, you need to format it carefully. You have to code up the image as a set of one band images stacked vertically, you need to set `page-height` to the height of each frame, and you have to attach some XML to `image-description` saying what each frame means. It's a bit hairy :(

I tried tiffinfo. It returned information of the 12 pyramids. I did not see the information of the 2 concatenated tiffs. Does it means 2 images are merged into one? It seems to contradict  with what I got in the python. The results did have 6 bands.

**Comment by jcupitt** [comment]
You need to make pyramids with `subifd=True` for QuPath -- it doesn't recognize ordinary image pyramids.

If you just bandjoin, you'll see this:

```
john@kiwi:~/pics$ vips bandjoin "k2.jpg k2.jpg" x.tif
john@kiwi:~/pics$ tiffinfo x.tif
TIFF Directory at offset 0x10fe008 (17817608)
  Image Width: 1450 Image Length: 2048
  Resolution: 72.009, 72.009 pixels/inch
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Extra Samples: 3<unassoc-alpha, unassoc-alpha, unassoc-alpha>
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 6
  Rows/Strip: 128
  Planar Configuration: single image plane
```

ie. an RGB image with six samples per pixel. For QuPath, you need to split the bands, as in that example code I linked.

**Comment by gmh1033286494** [comment]
I find it wired that result image is almost the same size as one of my 9 images instead of being 9 times the size of one image.
I think I need to understand why this happened and then try to figure out how to show my result. I am wondering if I am doing this right. I check my result image before I use **write_to_file** . There are indeed 9 bands in it. Somehow when I read the output read the output image, it became 6 bands. 
My goal is to put my 5 fixed image channels(cy5.5 dapi spgold spgreen spred) together with my 4 registered moving image channels( cy5.5 spgold spgreen spred). And, hopefully the result could be read by Halo and still there are 9 channels. 

**This is my code.**
_concatenate_list = []
concatenate_list.append(fixed_image_cy)
concatenate_list.append(fixed_image_dapi)
concatenate_list.append(fixed_image_Spgold)
concatenate_list.append(fixed_image_Spgreen)
concatenate_list.append(fixed_image_Spred)
concatenate_list.append(new_moving_spgold)
concatenate_list.append(new_moving_spgreen)
concatenate_list.append(new_moving_spred)
concatenate_list.append(new_moving_cy)

result_image_ = concatenate_list[0].bandjoin(concatenate_list[1:])

**The tiffinfo:**
TIFF Directory at offset 0x13aebee8 (330219240)
  Image Width: 84092 Image Length: 122880
  Tile Width: 128 Tile Length: 128
  Resolution: 36522.1, 36522.1 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: JPEG
  Photometric Interpretation: YCbCr
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Planar Configuration: single image plane
  Reference Black/White:
     0:     0   255
     1:   128   255
     2:   128   255
  SubIFD Offsets: 425773942 453089462 460833500 462971032 463539652 463691102 463735090 463748726 463753984 463756412
  ImageDescription: 
83456x122880 (256x256) JPEG/RGB Q=80|Date = 04/11/2020|Time = 11:44:49|MPP = 0.27381|3dh_PixelSizeX = 0.27381|3dh_PixelSizeY = 0.27381|3dh_Filter = Cy5.5|3dh_Profile = Default|3dh_ScannerHardwareID = PMIDI20C0301
  JPEG Tables: (142 bytes)

result_image.write_to_file("result_image.tiff", pyramid=True, tile=True, compression="jpeg",subifd=True)_

**Comment by jcupitt** [comment]
Don't use bandjoin! QuPath is weird -- you need to split the bands and code them separately. Have a look at that sample code I linked above.

**Comment by gmh1033286494** [comment]
Hi, @jcupitt. Merry Christmas!
I have been researching and trying the sample code. 
I find that I have a very similar situation with what described in this question: [https://github.com/libvips/libvips/issues/1253].
**That is to get a overlay image can be read by HALO or QUPATH**
So I tried this code 

_#!/usr/bin/python3

import sys
import pyvips

images = [pyvips.Image.new_from_file(filename, access="sequential") for filename in sys.argv[1:]]
out = pyvips.Image.arrayjoin(images, across=1)
out.set_type(pyvips.GValue.gint_type, "page-height", images[0].height)
print("writing x.tif ...")
out.write_to_file("x.tif")_

It seems that I did get a multi-page document. But Qupath still can not read it. So I tried to combine this with the sample code in #222 (comment).

_image_height = fixed_image_dapi.height
image_bands = fixed_image_dapi.bands

out = pyvips.Image.arrayjoin([fixed_image_dapi,fixed_image_Spgold], across=1)
out.set_type(pyvips.GValue.gint_type, "page-height", fixed_image_dapi.height)
out.set_type(pyvips.GValue.gstr_type, "image-description", 
f"""<?xml version="1.0" encoding="UTF-8"?>
<OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
    <Image ID="Image:0">
        
        <Pixels DimensionOrder="XYCZT"
                ID="Pixels:0"
                SizeC="{image_bands}"
                SizeT="1"
                SizeX="{fixed_image_dapi.width}"
                SizeY="{image_height}"
                SizeZ="1"
                Type="uint8">
        </Pixels>
    </Image>
</OME>""")
print("writing x.tif ...")
#out.write_to_file("x.tif", pyramid=True, tile=True, compression="jpeg")
#out.write_to_file("x_non_pyr_tiled.tif")
out.tiffsave('x_t2.tif', compression="jpeg", tile=True,
            tile_width=512, tile_height=512,
            pyramid=True, subifd=True)_

Qupath still can not read it. 
Sorry for my confusious understanding towards overlay tiff image.

**Comment by jcupitt** [comment]
It's working for me. I ran the sample program like this (the JPG is a pic of one of my kids):

```
./convert.py ~/pics/nina.jpg x.tif
```

To make:

http://www.rollthepotato.net/~john/x.tif

It loads into QuPath like this:

```
$ ./QuPath-0.2.3 ~/try/x.tif
...
12:38:15.781 [JavaFX Application Thread] [WARN ] loci.formats.FormatHandler - Using TiffReader to determine the number of planes.
12:38:15.942 [region-store-1] [WARN ] loci.formats.FormatHandler - SamplesPerPixel mismatch: OME=-1, TIFF=1
12:38:15.942 [region-store-1] [WARN ] loci.formats.FormatHandler - Image ID 'Image:0': missing plane #0
12:38:15.942 [region-store-1] [WARN ] loci.formats.FormatHandler - Image ID 'Image:0': missing plane #1
12:38:15.942 [region-store-1] [WARN ] loci.formats.FormatHandler - Image ID 'Image:0': missing plane #2
12:38:15.942 [region-store-1] [WARN ] loci.formats.FormatHandler - Using TiffReader to determine the number of planes.
12:38:16.272 [JavaFX Application Thread] [INFO ] qupath.lib.gui.viewer.QuPathViewer - Image data set to ImageData: Not set, x.tif
```

And pops up a dialog warning about stuff not being set, but displays fine. I tried a six-band image and that worked too.

**Comment by gmh1033286494** [comment]
It worked. It seems that after affined, the moving image get bigger. It could not match with the fixed image. So I reshape them into the same size.

**Comment by Yuvi-416** [comment]
> It worked. It seems that after affined, the moving image get bigger. It could not match with the fixed image. So I reshape them into the same size.

Hi there,
I am also suffering from a same problem. I am passing the same code as it is here #222 to convert any image into .tif. 
It worked and generate the pyramid style image but the generated first or bigger image output is repeated of each tiles while seeing it at ImageJ.
Input file and generated outputs are as follows:
This is a input image:
![1](https://user-images.githubusercontent.com/65744819/105647652-087aab00-5e9f-11eb-946d-5f2ab0012498.jpg)
And these are the screenshot of generated image (I can't upload .tif image bcoz github is not letting me to upload it)
![Screenshot from 2021-01-24 23-51-49](https://user-images.githubusercontent.com/65744819/105647722-79ba5e00-5e9f-11eb-8c17-c2818972de86.png)
![Screenshot from 2021-01-24 23-52-07](https://user-images.githubusercontent.com/65744819/105647723-7b842180-5e9f-11eb-81bd-12444ca314f0.png)
Can you say how did u solve your problem?
And do you know why the generated output has a separated channel (RGB)? Do you have any idea how to get a color image)

**Comment by gmh1033286494** [comment]
@Yuvi-416 

If you are using ImageJ，and your output is a multi-page image，I don't think you can tile and pyramid the image. I got the same result with tiled image in Fiji. HALO can't read tiled multi-page image，either. I don't know why. 
My solution is to tiffsave only with compression = 'lzw', or 'jpeg' if you don't mind the loss. 

For your RGB question I think it is because when you turn your input into multi-page image, the code spilt each RGB channels(bands) into a grey scale channels and the put them together into a multi-page image. My input images are all grey scale image with only one channel.

**Comment by apal6** [comment]
Hi, 

I am trying to do something similar. I have four RGB files that I want to combine to create a multi-channel tif file. I basically concatenated then to get a multi-dimentional file. However, when I tried loading it on ImageJ, it says unsupported file format. Not sure what I am doing is correct or not but this is what I tried. The size of each image was about 4gb and after concatenating, its around 18gb. My goal is to perfrom segmentation on this multi-channel image. Code I used:

`concat = np.concatenate((dapi_, CD3_, tcf7_, CD8_), axis=2)
np.array(concat).shape
output: (1, 47745, 136532, 3)`

`skio.imsave("/test.tif",concat)`

Thank you for your advice!
